File: frontend/demo/component/stepper/stepper-custom.ts
import '@vaadin/stepper';
import '@vaadin/icon';
import '@vaadin/icons';
import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('stepper-custom')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-stepper>
        <vaadin-step state="completed">
          <vaadin-icon icon="vaadin:user" slot="prefix"></vaadin-icon>
          <span>Account</span>
        </vaadin-step>
        <vaadin-step state="active">
          <vaadin-icon icon="vaadin:envelope" slot="prefix"></vaadin-icon>
          <span>Email verification</span>
        </vaadin-step>
        <vaadin-step>
          <vaadin-icon icon="vaadin:lock" slot="prefix"></vaadin-icon>
          <span>Security</span>
        </vaadin-step>
        <vaadin-step>
          <vaadin-icon icon="vaadin:check" slot="prefix"></vaadin-icon>
          <span>Complete</span>
        </vaadin-step>
      </vaadin-stepper>
      <!-- end::snippet[] -->
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/stepper/StepperCustom.java
package com.vaadin.demo.component.stepper;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.icon.Icon;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.stepper.Step;
import com.vaadin.flow.component.stepper.Stepper;
import com.vaadin.flow.router.Route;

@Route("stepper-custom")
public class StepperCustom extends Div {

    public StepperCustom() {
        // tag::snippet[]
        Stepper stepper = new Stepper();
        
        Step step1 = new Step();
        step1.add(new Icon(VaadinIcon.USER), new Span("Account"));
        step1.setState(Step.State.COMPLETED);
        
        Step step2 = new Step();
        step2.add(new Icon(VaadinIcon.ENVELOPE), new Span("Email verification"));
        step2.setState(Step.State.ACTIVE);
        
        Step step3 = new Step();
        step3.add(new Icon(VaadinIcon.LOCK), new Span("Security"));
        
        Step step4 = new Step();
        step4.add(new Icon(VaadinIcon.CHECK), new Span("Complete"));
        
        stepper.add(step1, step2, step3, step4);
        // end::snippet[]

        add(stepper);
    }

    public static class Exporter extends DemoExporter<StepperCustom> { // hidden-source-line
    } // hidden-source-line
}